x-pack/plugins/saved_objects_tagging/server/tags/tags_client.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract, SavedObject } from 'src/core/server';
+import { Tag, TagAttributes, ITagsClient } from '../../common/types';
+import { tagSavedObjectTypeName } from '../../common/constants';
+import { TagValidationError } from './errors';
+import { validateTag } from './validate_tag';
+
+type TagSavedObject = SavedObject<TagAttributes>;
+
+interface TagsClientOptions {
+  client: SavedObjectsClientContract;
+}
+
+const savedObjectToTag = (savedObject: TagSavedObject): Tag => {
+  return {
+    id: savedObject.id,
+    ...savedObject.attributes,
+  };
+};
+
+export class TagsClient implements ITagsClient {
+  private readonly soClient: SavedObjectsClientContract;
+  private readonly type = tagSavedObjectTypeName;
+
+  constructor({ client }: TagsClientOptions) {
+    this.soClient = client;
+  }
+
+  public async create(attributes: TagAttributes) {
+    const validation = validateTag(attributes);
+    if (!validation.valid) {
+      throw new TagValidationError('Error validating tag attributes', validation);
+    }
+    const raw = await this.soClient.create<TagAttributes>(this.type, attributes);
+    return savedObjectToTag(raw);
+  }
+
+  public async update(id: string, attributes: TagAttributes) {
+    const validation = validateTag(attributes);
+    if (!validation.valid) {
+      throw new TagValidationError('Error validating tag attributes', validation);
+    }
+    const raw = await this.soClient.update<TagAttributes>(this.type, id, attributes);
+    return savedObjectToTag(raw as TagSavedObject); // all attributes are updated, this is not a partial
+  }
+
+  public async get(id: string) {
+    const raw = await this.soClient.get<TagAttributes>(this.type, id);
+    return savedObjectToTag(raw);
+  }
+
+  public async getAll() {
+    const result = await this.soClient.find<TagAttributes>({
+      type: this.type,
+      perPage: 10000,
+    });
+
+    return result.saved_objects.map(savedObjectToTag);
+  }
+
+  public async delete(id: string) {
```

Comment:
Well, that's a tricky one. I discussed it a bit with @kobelb, and that's a similar issue of what security did with the space client (using the space client, deleting a space removes all objects within the space, even if it is possible to use the SO API to remove the space directly, which will NOT have this 'side' effect).

Ideally, we would not allow manipulating tags with the SO APIs. However the `hidden` solution is not acceptable here, as setting `tags` as hidden would have consequences on export/import/share_to_space actions.

We need a 'new' way to opt-out of SO APIs while preserving the capability for the type to be exportable / importable and to be included in any additional actions using the 'default repository'. But that's way beyond this PR's scope, so I think we will have to live with this limitation for now.

Reply:
We already have `management.importableAndExportable` so I think `hidden` should only affect the API so that you can have `hidden: true` and `importableAndExportable: true` but I agree it's out of the scope of this PR, I created https://github.com/elastic/kibana/issues/82027

Reply:
SOM is using a 'standard' SO client, so I think `hidden: true` will forbid to use the type at all during SO export/import (but did not test tbh)

Reply:
yeah I did a quick test, hidden types that are `importableAndExportable: true` still show up in the management UI, but always show "0" results and when exporting no objects are included.